zonemaster/zonemaster-backend
# Custom error message

## Purpose

Provide more human friendly error messages than the ones generated by the JSON validator.

## Context

#789

## Changes



The json validation error are mapped to a set of custom defined messages by using pattern matching on the `path`. The error is replaced with the one corresponding to the first `path` that matches. This is because the module we use for validation does not allow custom error messages (and internalization, in anticipation of #823 )  to be easily defined as it is not a specification of the JSON schema. See comment https://github.com/zonemaster/zonemaster-backend/pull/795#discussion_r667109704 and the following one for additional context .

Custom error messages implemented (with source of message):
* Domain names (`/domain`, `/nameserver/*/ns`)
  - The domain name is not a valid IDNA string and cannot be converted to an A-label (`_check_domain`)
  - The domain name contains non-ascii characters and IDNA is not installed (`_check_domain` and `json_schemas`)
  - The domain name character(s) are not supported (`_check_domain`)
* IPs ( `/nameserver/*/ip`)
  - Invalid IP address (`validate_syntax` for parsing check, `json_schemas` for pattern check)
* DS key tag (`/ds_info/*/keytag`)
  - Keytag should be a positive integer (`json_schemas` for type check and min(0) check)
* DS digest data (`/ds_info/*/digest`)
  - Invalid digest format (`json_schemas` for pattern check (length and characters allowed))
* Profile (`/profile`)
  - Unknown profile (`validate_syntax`)
* Language (`/language`)
  - Invalid language tag format
  - Language string not unique
  - Unkown language string
* Generic
  - Missing property

## How to test this PR
`curl -sH "Content-Type: application/json" --data '{"method":"start_domain_test","params":{"domain":"toto.com", "nameservers":[{"ip":"a"}], "ds_info":[{"algorithm":-1,"digest":"abc","digtype":-1,"keytag":"abc"}]},"id":1,"jsonrpc":"2.0"}' localhost:5000/api  | jq`
```json
{
  "id": 1,
  "error": {
    "data": [
      {
        "message": "Algorithm should be a positive integer",
        "path": "/ds_info/0/algorithm"
      },
      {
        "path": "/ds_info/0/digest",
        "message": "Invalid digest format"
      },
      {
        "path": "/ds_info/0/digtype",
        "message": "Digest type should be a positive integer"
      },
      {
        "path": "/ds_info/0/keytag",
        "message": "Keytag should be a positive integer"
      },
      {
        "path": "/nameservers/0/ip",
        "message": "Invalid IP address"
      },
      {
        "path": "/nameservers/0/ns",
        "message": "Missing property"
      },
      {
        "path": "/nameservers/0/ns",
        "message": "Domain name required"
      },
      {
        "message": "Invalid IP address",
        "path": "/nameservers/0/ip"
      }
    ],
    "code": "-32602",
    "message": "Invalid method parameter(s)."
  },
  "jsonrpc": "2.0"
}
```
